### PR TITLE
Add `build_request` for `Client`

### LINKED ISSUE #174: Create a Client.build_request() method
See: https://github.com/encode/httpx/issues#issuecomment-515950955

### PR DESCRIPTION
**Pull request by dobrovolsky**
Using a Client instance to make build requests
```python
>>> client = httpx.Client()
>>> r = client.build_request("GET", "https://example.com")
>>> r
<AsyncRequest('GET', 'https://example.com')>
>>> client.send(r)
<Response [200 OK]>
```
And async client
```python
>>> import httpx
>>> client = httpx.AsyncClient()
>>> r = client.build_request("GET", "https://example.com")
>>> r
<AsyncRequest('GET', 'https://example.com')>
>>> client.send(r)
<coroutine object AsyncClient.send at 0x7fbc588f7560>
```

Closes #174

### CODE REVIEW

File: httpx/client.py [hunk 1]
```diff
@@ -237,6 +237,33 @@ def _get_auth_middleware(
 
         return None
 
+    def build_request(
```
Comment (cansarigol) [inline httpx/client.py]: is `build_request` func's position correct in BaseClient ?

Comment (cansarigol) [reply httpx/client.py]: I asked to understand if BaseClient is an abstract class for all custom client classes. so far it is used to narrow `Client` and `AsyncClient` code down.

File: docs/advanced.md [hunk 2]
```diff
@@ -57,6 +57,28 @@ dispatch = httpx.WSGIDispatch(app=app, remote_addr="1.2.3.4")
 client = httpx.Client(dispatch=dispatch)
 ```
 
+## Build Request
+
+Using a Client instance to make build requests
```
Comment (sethmlarson) [inline docs/advanced.md]: We should probably show the use-case here such as being able to modify the request in ways you usually can't with our standard interface. An example of this is making an `OPTIONS *` request for CORS:

```python
client = httpx.Client()
req = client.build_request("OPTIONS", "https://example.com")
req.url.full_path = "*"
resp = client.send(req)
```

(I don't think you can *currently* do this but you'll be able to for sure after my proxy PR lands)

File: docs/advanced.md [hunk 3]
```diff
@@ -57,6 +57,28 @@ dispatch = httpx.WSGIDispatch(app=app, remote_addr="1.2.3.4")
...
+```python
+>>> client = httpx.Client()
+>>> r = client.build_request("GET", "https://example.com")
+>>> r
+<AsyncRequest('GET', 'https://example.com')>
+>>> client.send(r)
+<Response [200 OK]>
+```
+And async client
```
Comment (sethmlarson) [inline docs/advanced.md]: Let's not have a duplicate demonstration for the AsyncClient, client and async client have the same interface, only show where they differ.

File: httpx/client.py [hunk 4]
```diff
@@ -153,7 +153,7 @@ def merge_headers(
             return merged_headers
         return headers
 
-    async def send(
+    async def _send(
```
Comment (sethmlarson) [inline httpx/client.py]: I'm not sure I understand the reasoning of this change, can we not keep the `send()` methods the same as they were previously?

File: tests/client/test_async_client.py [hunk 5]
```diff
@@ -14,6 +14,18 @@
     assert repr(response) == "<Response [200 OK]>"
 
 
+async def test_build_request(server, backend):
+    url = server.url
+    async with httpx.AsyncClient(backend=backend) as client:
+        request = client.build_request("GET", url)
+        response = await client.send(request)
```
Comment (sethmlarson) [inline tests/client/test_async_client.py]: Should probably modify the request in some way before sending it and then verify the request is modified in the result.

File: tests/client/test_client.py [hunk 6]
```diff
@@ -19,6 +19,23 @@ def test_get(server):
     assert repr(response) == "<Response [200 OK]>"
 
 
+def test_build_request(server):
+    url = server.url
+    with httpx.Client() as http:
+        request = http.build_request("GET", url)
```
Comment (sethmlarson) [inline tests/client/test_client.py]: Same as above review comment.

File: httpx/client.py (hunk 1 above)
Comment (dobrovolsky) [reply httpx/client.py]: I think so
`Client` and `AsyncClient` build Request in same way

File: httpx/client.py (hunk 4 above)
Comment (dobrovolsky) [reply httpx/client.py]: we need some internal method for sending request
also we need different method with different signature 
`send() -> Response` and `send() -> AsyncResponse`

File: tests/client/test_async_client.py (hunk 5 above)
Comment (dobrovolsky) [reply tests/client/test_async_client.py]: do you have any suggestion for testing?

File: docs/advanced.md (hunk 2 above)
Comment (dobrovolsky) [reply docs/advanced.md]: In your example we can't change `full_path`  because it's an attribute
Do you have any suggestions?

Comment (sethmlarson) [reply docs/advanced.md]: You can't change it now but once another PR is merged you will be able to.
You shouldn't create a test case with it, but you can add it to documentation. :)

File: httpx/client.py (hunk 4 above)
Comment (sethmlarson) [reply httpx/client.py]: You're right! I take back this comment, although this method might need a better name?

File: tests/client/test_async_client.py (hunk 5 above)
Comment (sethmlarson) [reply tests/client/test_async_client.py]: Maybe adding a header, just something that wouldn't be done by default.

**Comment by florimondmanca** [comment]
I think it's a shame that `.send()` and `.dispatch` are already taken on the client, to be honest. In terms of public API, `response = client.send(request)` is what makes most sense even from a Requests compatibility point of view.

I'm thinking we could very well define a synchronous `def .send()` on `Client` that calls the asynchronous `super().send` coroutine, no?

Mypy would complain, but maybe that's what I would expect as a user: `Client` has a sync `.send()`, while `AsyncClient` has an `async .send()`.

See:

```python
async with httpx.AsyncClient() as client:
    request = client.build_request("GET", "https://example.com")
    response = await client.send(request)

with httpx.Client() as client:
    request = client.build_request("GET", "https://example.com")
    response = client.send(request)
```

That way we don't have to throw away the existing `.send()` verb that's used throughout the client, dispatchers and connections.

(Also it's actually the interface we already advertise in the [Developer Interface](https://www.encode.io/httpx/api/#client), but currently `.send()` is always async. Which means the docs are currently wrong, actually.)

Thoughts?

File: tests/conftest.py [hunk 7]
```diff
@@ -111,6 +113,25 @@ def backend(request):
     await send({"type": "http.response.body", "body": body})
 
 
+async def echo_headers(scope, receive, send):
+    body: bytes = b""
+
+    more_body = scope.get("headers", [])
+    for h in more_body:
```
Comment (florimondmanca) [inline tests/conftest.py]: Can we rephrase this to…?

```python
    for name, value in scope.get("headers", []):
```

File: docs/api.md [hunk 8]
```diff
@@ -38,6 +39,7 @@
 * `def .delete(url, [data], [json], [params], [headers], [cookies], [auth], [stream], [allow_redirects], [verify], [cert], [timeout])`
 * `def .request(method, url, [data], [params], [headers], [cookies], [auth], [stream], [allow_redirects], [verify], [cert], [timeout])`
 * `def .send(request, [stream], [allow_redirects], [verify], [cert], [timeout])`
+* `def .build_request(method, url, [data], [files], [json], [params], [headers], [cookies])`
```
Comment (florimondmanca) [inline docs/api.md]: For consistency, let's move this to below `.request()` and above `.send()`?

File: tests/conftest.py [hunk 9]
```diff
@@ -111,6 +113,23 @@ def backend(request):
     await send({"type": "http.response.body", "body": body})
 
 
+async def echo_headers(scope, receive, send):
+    body: bytes = b""
+
+    for name, value in scope.get("headers", []):
+        value = f"{name.capitalize().decode()}: {value.decode()}\n"
+        body += value.encode()
```
Comment (yeraydiazdiaz) [inline tests/conftest.py]: Nit: Maybe it'd be more ergonomic to return the headers as a JSON object? I think it would make the assertions more readable.

Comment (dobrovolsky) [reply tests/conftest.py]: Nice hint, thanks

File: tests/client/test_client.py [hunk 10]
```diff
@@ -19,6 +19,24 @@ def test_get(server):
...
+
+    with httpx.Client() as http:
+        request = http.build_request("GET", url)
+        request.headers.update(headers)
+        response = http.send(request)
+
+    assert response.status_code == 200
+    assert response.url == url
+
+    response_data = response.json()
```
Comment (sethmlarson) [inline tests/client/test_client.py]: Since we're only asserting one value we can simplify our asserts like this: `assert response.json()["custom-header"] == "value"`

**Comment by sethmlarson** [comment]
Thanks @dobrovolsky! :clap:
